### PR TITLE
Update cpp-app template to use ReactNativeAppBuilder

### LINKED ISSUE #12991: Create public OSS new app templates for React Native Fabric
The tooling and templates necessary for end developers to create their own Win32 app which hosts a React Native Fabric Island. This will need to conform to the expectations and norms of the existing RN app development ecosystem. Without this we will not be able to keep or attract external developers to React Native for Windows after the Paper renderer is deprecated.
```[tasklist]
### Tasks
- [ ] https://github.com/microsoft/react-native-windows/pull/13875
```

### PR DESCRIPTION
**Pull request by sharath2727**
### Why
Simplify the default template code for creating the fabric app

Resolves #12991 

### What
Updated the cpp-app template to use ReactNativeAppBuilder

## Changelog
Should this change be included in the release notes: yes
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/microsoft/react-native-windows/pull/13875)

### CODE REVIEW

File: vnext/templates/cpp-app/windows/MyApp/MyApp.cpp [hunk 1]
```diff
@@ -20,147 +20,61 @@ struct CompReactPackageProvider
...
-
-  host.PackageProviders().Append(winrt::make<CompReactPackageProvider>());
-
+  // DebugBundlePath is used when loading from a bundle server such as metro. If this parameter
+  // is not provided fallback to a combination of JavaScriptBundleFile and BundleRootPath
+  auto reactInstanceSettingsBuilder {
+    winrt::Microsoft::ReactNative::ReactInstanceSettingsBuilder()
+        .DebugBundlePath(L"index")
+        .JavaScriptBundleFile(L"index.windows")
+        .BundleRootPath(appDirectory)
```
Comment (jonthysell) [inline vnext/templates/cpp-app/windows/MyApp/MyApp.cpp:46]: this needs to maintain the `file://` prefix from before

File: vnext/templates/cpp-app/windows/MyApp/MyApp.cpp [hunk 2]
```diff
@@ -20,147 +20,61 @@ struct CompReactPackageProvider
...
-      winrt::Microsoft::UI::Windowing::OverlappedPresenterState::Minimized) {
-    winrt::Microsoft::ReactNative::LayoutConstraints constraints;
-    constraints.MaximumSize = constraints.MinimumSize = size;
-    rootView.Arrange(constraints, {0,0});
-  }
-}
+  // Enable per monitor DPI scaling
+  SetProcessDpiAwarenessContext(DPI_AWARENESS_CONTEXT_PER_MONITOR_AWARE_V2);
+
+  // Create a top-level window.
```
Comment (jonthysell) [inline vnext/templates/cpp-app/windows/MyApp/MyApp.cpp:30]: While this is an *option* with the builder, I don't want it in the template. It's fine in the sample app, but for the template i expect people to use the window we give them. and then have the title/resize specified later

File: vnext/templates/cpp-app/windows/MyApp/MyApp.cpp [hunk 3]
```diff
@@ -20,147 +20,61 @@ struct CompReactPackageProvider
...
-  auto scaleFactor = ScaleFactor(hwnd);
+  winrt::Microsoft::ReactNative::ReactViewOptions viewOptions;
+  viewOptions.ComponentName(mainComponentName);
 
-  auto host = CreateReactNativeHost(hwnd, compositor);
+  // Initialize and Manage the ReactNativeHost
+  auto reactNativeAppBuilder{winrt::Microsoft::ReactNative::ReactNativeAppBuilder()
+                                 .AddPackageProviders(packageProviders)
+                                 .SetReactInstanceSettings(reactInstanceSettingsBuilder.ReactInstanceSettings())
+                                 .SetAppWindow(window)
```
Comment (jonthysell) [inline vnext/templates/cpp-app/windows/MyApp/MyApp.cpp:73]: As per the above, don't set the window here.

File: vnext/templates/cpp-app/windows/MyApp/MyApp.cpp [hunk 4]
```diff
@@ -20,147 +20,61 @@ struct CompReactPackageProvider
...
+                                 .SetReactViewOptions(viewOptions)};
 
-  // Start the react-native instance, which will create a JavaScript runtime and load the applications bundle
-  host.ReloadInstance();
+  // Start the react-native instance by creating a javascript runtime and load the bundle.
+  auto reactNativeWin32App{reactNativeAppBuilder.Build()};
```
Comment (jonthysell) [inline vnext/templates/cpp-app/windows/MyApp/MyApp.cpp:77]: after building the app, let's use the window getter and set the title / size here.

File: vnext/templates/cpp-app/windows/MyApp/MyApp.cpp (hunk 1 above)
Comment (sharath2727) [reply vnext/templates/cpp-app/windows/MyApp/MyApp.cpp:46]: Ah actually this is taken care of in the reactnativeappbuilder

File: vnext/templates/cpp-app/windows/MyApp/MyApp.cpp (hunk 2 above)
Comment (sharath2727) [reply vnext/templates/cpp-app/windows/MyApp/MyApp.cpp:30]: done.

File: vnext/templates/cpp-app/windows/MyApp/MyApp.cpp (hunk 3 above)
Comment (sharath2727) [reply vnext/templates/cpp-app/windows/MyApp/MyApp.cpp:73]: done.

File: vnext/templates/cpp-app/windows/MyApp/MyApp.cpp (hunk 4 above)
Comment (sharath2727) [reply vnext/templates/cpp-app/windows/MyApp/MyApp.cpp:77]: done.

File: vnext/templates/cpp-app/windows/MyApp/MyApp.cpp [hunk 5]
```diff
@@ -20,147 +20,58 @@ struct CompReactPackageProvider
...
-
+  // DebugBundlePath is used when loading from a bundle server such as metro. If this parameter
+  // is not provided fallback to a combination of JavaScriptBundleFile and BundleRootPath
+  auto reactInstanceSettingsBuilder {
+    winrt::Microsoft::ReactNative::ReactInstanceSettingsBuilder()
 #if BUNDLE
-  host.InstanceSettings().JavaScriptBundleFile(L"index.windows");
-  host.InstanceSettings().BundleRootPath(std::wstring(L"file://").append(appDirectory).append(L"\\Bundle\\").c_str());
-  host.InstanceSettings().UseFastRefresh(false);
+        .BundleRootPath(appDirectory)
```
Comment (jonthysell) [inline vnext/templates/cpp-app/windows/MyApp/MyApp.cpp:39]:
```suggestion
        .BundleRootPath(std::wstring(L"file://").append(appDirectory).append(L"\\Bundle\\").c_str())
```

**Review by jonthysell — APPROVED** [review]
Looks great, just need to fix the bundle path (without the "file://" release builds won't load the bundle properly)

File: vnext/templates/cpp-app/windows/MyApp/MyApp.cpp (hunk 5 above)
Comment (jonthysell) [reply vnext/templates/cpp-app/windows/MyApp/MyApp.cpp:39]: I now see you put it in the `ReactInstanceSettingsBuilder`. End developers may want to put other urls in there, or customize the path in another way. We shouldn't force/hide adding "file://" and "\\Bundle" in the settings builder.

**Review by jonthysell — APPROVED** [review]

**Pull request merged**